### PR TITLE
drivers: ipm driver based on stm32 HSEM

### PR DESCRIPTION
**Pull request by cameled**
Some STM32 SOC, like stm32h745 and stm32h747 doesn't have IPCC.
Provide an STM32 HSEM based ipm driver for these SOC.

### CODE REVIEW

**Comment by erwango** [comment]
@cameled Great thanks for this contribution! 
To ease the review, is there a way we can test this ?
Ideally providing a way to build and test this driver in tree (so that at least this could be built by CI) would be a great addition.
Maybe a very basic sample that could be flashed on both cores would suffice. Anyway, this won't be a blocker as dual core testing is a particular case.

**Comment by cameled** [comment]
> @cameled Great thanks for this contribution!
> To ease the review, is there a way we can test this ?
> Ideally providing a way to build and test this driver in tree (so that at least this could be built by CI) would be a great addition.
> Maybe a very basic sample that could be flashed on both cores would suffice. Anyway, this won't be a blocker as dual core testing is a particular case.

**Comment by erwango** [comment]
> @cameled cameled closed this 42 seconds ago

@cameled, I'm sorry you closed this PR. Hope this is not definitive.
Please note that my request about adding a test is not blocking due to the specific dual core condition.
This can be done as a later steps and I'm sure we'll find help in the community to add it.

**Comment by cameled** [comment]
Sorry, I think it close the comment, but it close the whole pull request.

**Comment by cameled** [comment]
I'm preparing an simple test example now.
I have a stm32h747i_disco board, so example will work on it.

File: dts/arm/st/h7/stm32h7.dtsi [hunk 1]
```diff
@@ -737,6 +737,15 @@
 			#io-channel-cells = <1>;
 		};
 
+		mailbox: mailbox@58026400 {
+			compatible = "st,stm32-hsem-mailbox";
+			reg = <0x58026400 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_AHB4 0x02000000>;
+			interrupts = <125 0>;
```
Comment (ABOSTM) [inline dts/arm/st/h7/stm32h7.dtsi:744]: As interrupt 125 is valid for Cortex M7 only, could it be moved to _m7 specific file ?

**Review by ABOSTM — CHANGES_REQUESTED** [review]
@cameled, great job thanks.
2 minor remarks otherwise, LGTM

File: dts/arm/st/h7/stm32h7.dtsi [hunk 2]
```diff
@@ -737,6 +737,15 @@
 			#io-channel-cells = <1>;
 		};
 
+		mailbox: mailbox@58026400 {
```
[the remark this answers is not in the document]
Comment (cameled) [reply dts/arm/st/h7/stm32h7.dtsi:740]: OK, I will move it to stm32h7_dualcore.dtsi

File: dts/arm/st/h7/stm32h7.dtsi (hunk 1 above)
Comment (cameled) [reply dts/arm/st/h7/stm32h7.dtsi:744]: OK, I will try to move it to _m7, same like _m4.

File: drivers/ipm/ipm_stm32_hsem.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+static const struct stm32_hsem_mailbox_config stm32_hsem_mailbox_0_config = {
+	.irq_config_func = stm32_hsem_mailbox_irq_config_func,
+	.pclken = {
+		.bus = DT_INST_CLOCKS_CELL(0, bus),
+		.enr = DT_INST_CLOCKS_CELL(0, bits)
+	},
+};
+
+DEVICE_DT_INST_DEFINE(0,
```
Comment (mbolivar-nordic) [inline drivers/ipm/ipm_stm32_hsem.c:192]: Is there a good reason why this does not respect `status = "disabled";` by using DT_INST_FOREACH_STATUS_OKAY?

Comment (cameled) [reply drivers/ipm/ipm_stm32_hsem.c:192]: STM32 HSEM has its own LL_HSEM(low-level HSEM) API provided by the hal_stm32 module. LL_HSEM API worked very well for the hardware semaphore mechanism.
The STM32 HSEM IPM driver I add here, Only pick up two semaphore ID from stm32_hsem.h to simulate a virtual mailbox device. 
So there will have only one instance.

Comment (mbolivar-nordic) [reply drivers/ipm/ipm_stm32_hsem.c:192]: Having a single instance does not excuse ignoring status 🙂. I think the default stance should be to respect it.

Comment (cameled) [reply drivers/ipm/ipm_stm32_hsem.c:192]: Sorry, misunderstanding the question, will add the 'status' check.

Comment (cameled) [reply drivers/ipm/ipm_stm32_hsem.c]: I have added the DT_INST_FOREACH_STATUS_OKAY.

File: boards/arm/stm32h747i_disco/Kconfig.defconfig [hunk 4]
```diff
@@ -12,6 +12,14 @@ config BOARD
 config STM32H7_DUAL_CORE
 	default y
 
+if IPM_STM32_HSEM
```
Comment (erwango) [inline boards/arm/stm32h747i_disco/Kconfig.defconfig:15]: Please also update board documentation

**Review by erwango — COMMENTED** [review]
Thanks again for the great work.

Ideally, we'd request to split in various commits:
- dtsi/bindings
- soc
- driver
- board
Though this is not blocking.

Thanks for sharing the sample. It would be nice to have it under `samples/boards/stm32/h7_dual_core` so people can easily test it and make sure the driver is compiled in CI.

EDIT: I've performed the review in comments instead of requesting changes, as I will be out for some time and I would not want to block merging.

**Comment by erwango** [comment]
@cameled Out of curiosity, are you planning to use openamp on top of this driver ?

**Comment by cameled** [comment]
> @cameled Out of curiosity, are you planning to use openamp on top of this driver ?
Yes, I'm preparing to use openamp.

**Comment by cameled** [comment]
It's my first-time to contribute code for an open source project.
If there have some unsuitable things, please let me know. 
Thanks a lot!

File: drivers/ipm/ipm_stm32_hsem.c [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+#if CONFIG_IPM_STM32_HSEM_CPU == HSEM_CPU1
+#define LL_HSEM_EnableIT_CIER       LL_HSEM_EnableIT_C1IER
+#define LL_HSEM_DisableIT_CIER      LL_HSEM_DisableIT_C1IER
+#define LL_HSEM_ClearFlag_CICR      LL_HSEM_ClearFlag_C1ICR
+#define LL_HSEM_IsActiveFlag_CISR   LL_HSEM_IsActiveFlag_C1ISR
+#else /* HSEM_CPU2 */
+#define LL_HSEM_EnableIT_CIER       LL_HSEM_EnableIT_C2IER
+#define LL_HSEM_DisableIT_CIER      LL_HSEM_DisableIT_C2IER
+#define LL_HSEM_ClearFlag_CICR      LL_HSEM_ClearFlag_C2ICR
+#define LL_HSEM_IsActiveFlag_CISR   LL_HSEM_IsActiveFlag_C2ISR
```
[the remark this answers is not in the document]
Comment (cameled) [reply drivers/ipm/ipm_stm32_hsem.c:31]: Yes, you are right. to avoid confusion with STM32 HSEM low-level API, I will change the redefined function name.

**Comment by erwango** [comment]
> It's my first-time to contribute code for an open source project.

And you're doing great!

> If there have some unsuitable things, please let me know.

Nothing is unsuitable, but please address the inline comments provided above.

File: boards/arm/stm32h747i_disco/Kconfig.defconfig (hunk 4 above)
Comment (cameled) [reply boards/arm/stm32h747i_disco/Kconfig.defconfig:15]: OK, I will update the board documentation.

**Comment by cameled** [comment]
@erwango The change will be sent Monday or Tuesday on next week.

**Review by mbolivar-nordic — APPROVED** [review]
Thanks for the update and for addressing my change request. Welcome to the project!

File: drivers/ipm/Kconfig [hunk 6]
```diff
@@ -112,6 +112,20 @@ config IPM_INTEL_ADSP
 	help
 	  Driver for the Host-DSP Mailbox Communication channel.
 
+config IPM_STM32_HSEM
+	bool "STM32 HSEM controller"
+	depends on STM32H7_DUAL_CORE
+	help
+		Driver for stm32 HSEM mailbox
```
[the remark this answers is not in the document]
Comment (cameled) [reply drivers/ipm/Kconfig:119]: Thanks.

File: drivers/ipm/Kconfig [hunk 7]
```diff
@@ -112,6 +112,20 @@ config IPM_INTEL_ADSP
...
+
+config IPM_STM32_HSEM_CPU
+	int "HSEM CPU ID"
+	default 1
+	range 1 2
+	depends on IPM_STM32_HSEM
+	help
+		use to define the CPU ID used by HSEM
```
[the remark this answers is not in the document]
Comment (cameled) [reply drivers/ipm/Kconfig:127]: Thanks.

**Comment by cameled** [comment]
Sample have some problem. create a issue to track it. #37395

**Review by ABOSTM — APPROVED** [review]
LGTM

**Pull request merged**